docs/manual/index-checker.tex

```diff
@@ -457,8 +457,119 @@ \section{Binary search indices\label{index-searchindex}}
+and \sunjavadoc{java/lang/String.html\#lastIndexOf-java.lang.String-}{String.lastIndexOf} return an index
+of a given substring within a given string, or -1 if no such substring exists.
+The index \<i> returned from \<receiver.indexOf(substring)> satisfies this property:
+\code{i==-1 || i>=0 \&\& i<=receiver.length() - substring.length()}.
+
+This tells us what Index Checker annotations can be applied to \<i>:
+\begin{itemize}
+\item
+\<i> is \<@GTENegativeOne>, because \code{i>=-1}.
+\item
+\<i> is \<@LTEqLengthOf("receiver")>, because \code{i<=receiver.length()}.
+\item
+\<i> is not \<@IndexOrLow("receiver")>, because for \code{receiver="", substring="", i=0},
+the property \code{i>=-1 \&\& i<receiver.length()} does not hold.
+\item
+\<i> is not \<@IndexOrHigh("receiver")>, because for \code{receiver="", substring="b", i=-1},
+the property \code{i>=0 \&\& i<=receiver.length()} does not hold.
+\item
+\<i> is not \<@LTLengthOf(value="receiver", offset="substring.length()-1")>,
+because for \code{receiver="", substring="abc", i=-1},
+the property \code{i+substring.length()-1<receiver.length()} does not hold.
+\end{itemize}
+The last annotation in the list above, \<@LTLengthOf(value="receiver", offset="substring.length()-1")>,
+is the correct and precise upper bound for all values of \<i> except -1.
+The offset expresses the fact that we can add \<substring.length()> to
+this index and still get a valid index for \<receiver>.
+That is useful for type-checking code, where we add the length of the substring
+to the index to get the rest of the string after the occurrence we found.
+However, the upper bound applies only after the index is explicitly checked not to be -1:
+
+\begin{Verbatim}
+ int i = receiver.indexOf(substring);
+ // i is @GTENegativeOne and @LTEqLengthOf("receiver")
+ // i is not @LTLengthOf(value="receiver", offset="substring.length()-1")
+ if (i != -1) {
+  // i is @NonNegative and @LTLengthOf(value="receiver", offset="substring.length()-1")
+  int j = i + substring.length();
+  // j is @IndexOrHigh("receiver")
+  return receiver.substring(j); // this call is safe
+ }
+\end{Verbatim}
+
+The property of the result of \<indexOf> cannot be expressed by any combination of lower- and upper-bound annotations.
+It's because the upper-bound annotations apply independently of the lower-bound annotations,
+but in this case, the upper bound \code{i<=receiver.length() - substring.length()} holds only if \code{i>=0}.
+Therefore, to express this property and make the example type-check without false positives,
+a new annotation such as
+\<@IndexOfIndexFor(value="receiver", offset="substring.length()-1")>
+%Alternaive:
+%\<@IndexOfSubstring\allowbreak(value="substring", in="receiver")>
+is necessary.
+
+This new annotation is similar to \<@LTLengthOf\allowbreak(value="receiver", offset="substring.length()-1")>,
+but explicitly allows the index to be -1 even if the upper bound would not allow it.
```

Comment:
I would say "even if the upper bound's offset would not allow it" or something along those lines, to help the reader recall the case above where the correct annotation is an `LTLengthOf` with an offset as long as the index is not -1.